### PR TITLE
BUG: File expansion when updating with Page Contents

### LINKED ISSUE #1897: File Size Increase from pypdf==3.8.1 to pypdf==3.9.0 when watermarking
I've noticed that the watermarking process in my job suddenly produces way bigger PDFs. In a specific example with [this PDF file](https://wiwin.de/wp-content/uploads/2023/03/Anlagebroschure-Vogtland-IV.pdf) I've noticed:

* Base is 3.5 MB
* Watermarking with `pypdf==3.8.1` produces a 3.7 MB file
* Watermarking with `pypdf==3.9.0` produces a 5.2 MB file

I'm not sure if the issue is with 3.8.1 or with 3.9.0.

## Code

```python
from pypdf import PdfReader, PdfWriter
import pypdf
from io import BytesIO
from fpdf import FPDF  # pip install fpdf2


def create_stamp_pdf() -> BytesIO:
    pdf = FPDF()
    pdf.add_page()
    pdf.set_font("helvetica", "B", 16)
    pdf.cell(40, 10, "Hello World!")
    byte_string = pdf.output()
    return BytesIO(byte_string)


template = PdfReader(create_stamp_pdf())
template_page = template.pages[0]
reader = PdfReader("template.pdf")

writer = PdfWriter()
for page in reader.pages:
    page.merge_page(template_page)
    writer.add_page(page)

for page in writer.pages:
    page.compress_content_streams()

with open(f"out-{pypdf.__version__}.pdf", "wb") as fp:
    writer.write(fp)
```

## Affected versions

It seems to be https://github.com/py-pdf/pypdf/commit/2d67c15d46a0488e619d3cdf24707feb2d92c224 that causes it. So pypdf>=3.9.0.

### PR DESCRIPTION
**Pull request by pubpub-zz**
closes #1897

### CODE REVIEW

**Comment by MartinThoma** [comment]
You have to install `fpdf2`. I can do that later and rebase your PR :-)

**Comment by MartinThoma** [comment]
Thank you for the quick fix :pray:

File: pypdf/_page.py [hunk 1]
```diff
@@ -920,6 +920,50 @@ def getContents(self) -> Optional[ContentStream]:  # deprecated
...
         return self.get_contents()
 
+    def replace_contents(self, content: Optional[ContentStream]) -> None:
+        """
+        Replace the page contents with the new content and nullify old objects
+        Args:
+            content : new content. if None delete the content field.
+        """
+        if isinstance(self.get(PG.CONTENTS, None), ArrayObject):
+            for o in self[PG.CONTENTS]:
```
Comment (MartinThoma) [inline pypdf/_page.py:930]:
```suggestion
            for o in self[PG.CONTENTS]:  # type: ignore[attr-defined]
```

File: pypdf/_page.py [hunk 2]
```diff
@@ -920,6 +920,50 @@ def getContents(self) -> Optional[ContentStream]:  # deprecated
...
+                try:
+                    self._objects[o.indirect_reference.idnum - 1] = NullObject()
```
Comment (MartinThoma) [inline pypdf/_page.py:932]:
```suggestion
                    self._objects[o.indirect_reference.idnum - 1] = NullObject()  # type: ignore
```

File: pypdf/_page.py [hunk 3]
```diff
@@ -920,6 +920,50 @@ def getContents(self) -> Optional[ContentStream]:  # deprecated
...
+                except AttributeError:
+                    pass
+        if content is None:
+            if PG.CONTENTS not in self:
+                return
+            else:
+                self.indirect_reference.pdf._objects[
```
Comment (MartinThoma) [inline pypdf/_page.py:939]: Mypy complains that `self.indirect_reference` might be none here. Could that actually happen?

Comment (pubpub-zz) [reply pypdf/_page.py:947]: Some cases for independant pages. I've added some code

**Comment by pubpub-zz** [comment]
Now coverage to be improve. give me a coupe of hours for dinner 😋

**Comment by MartinThoma** [comment]
Very nice! Thank you :-)

**Pull request merged**